### PR TITLE
Add optional validation metric for classification

### PR DESCRIPTION
**Pull request by danlamanna**
@brianhelba Do you want to use an enum for the metrics? Does the logic of this look good? How should `None` be handled?

### CODE REVIEW

File: isic_challenge_scoring/classification.py [hunk 1]
```diff
@@ -61,9 +62,19 @@ def __init__(
...
-        self.validation = metrics.balanced_multiclass_accuracy(
-            truth_probabilities, prediction_probabilities, truth_weights.validation_weight
-        )
+
+        if validation_metric == 'balanced_accuracy':
+            self.validation = metrics.balanced_multiclass_accuracy(
+                truth_probabilities, prediction_probabilities, truth_weights.validation_weight
+            )
+        elif validation_metric == 'ap':
+            self.validation = metrics.average_precision(
```
Comment (brianhelba) [inline isic_challenge_scoring/classification.py]: [`metrics.average_precision` takes arguments](https://github.com/ImageMarkup/isic-challenge-scoring/blob/master/isic_challenge_scoring/metrics.py#L206) of `pd.Series`, where each is the probabilities for a single category.

Here in the code, `truth_probabilities` is a `pd.DataFrame`, holding probabilities for all categories. You can reference `truth_probabilities[category]` to get a single category (see line 34 above).

I'm not sure why the type checking tests aren't catching this...

**Review by brianhelba — CHANGES_REQUESTED** [review]
There's a fundamental issue with the way that single-category metrics are being called. See the inline comment.

An enum is a good idea.

For the (default) case of no specified validation metric, perhaps just make [`Score.validation`](https://github.com/ImageMarkup/isic-challenge-scoring/blob/master/isic_challenge_scoring/types.py#L18) optional? I don't feel strongly if you have a more elegant approach.

Is it possible to add tests to exercise each of the new conditional cases (including the default)?

File: isic_challenge_scoring/classification.py (hunk 1 above)
Comment (danlamanna) [reply isic_challenge_scoring/classification.py]: For `auc` and `ap` should I just average the per category scores?

Comment (brianhelba) [reply isic_challenge_scoring/classification.py]: I think so. For the cases where `auc` and `ap` are used (past challenges), they are mostly single-category tasks, so the average will be a no-op. In one place (2017 Task 3), there are two categories, though it's up to our discretion what specifically to choose for the validation metric.

Be sure to [compute the macro average, not the micro average](https://datascience.stackexchange.com/a/16001), since that's what we do for most of the other averages.

File: isic_challenge_scoring/classification.py [hunk 2]
```diff
@@ -63,24 +63,29 @@ def __init__(
...
+
+                ) for category in categories])
+                self.validation = per_category_auc.mean(axis='index')
 
     @staticmethod
     def _category_score(
         truth_category_probabilities: pd.Series,
         prediction_category_probabilities: pd.Series,
-        truth_weights: pd.DataFrame,
+        truth_weights: pd.Series,
```
Comment (brianhelba) [inline isic_challenge_scoring/classification.py]: Good change. Can you split to another PR?

Comment (brianhelba) [reply isic_challenge_scoring/classification.py]: Also, maybe rename the parameter to `truth_score_weights`?

File: isic_challenge_scoring/classification.py [hunk 3]
```diff
@@ -63,24 +63,29 @@ def __init__(
...
-            )
-        elif validation_metric == 'ap':
-            self.validation = metrics.average_precision(
-                truth_probabilities, prediction_probabilities, truth_weights.validation_weight
-            )
-        elif validation_metric == 'auc':
-            self.validation = metrics.auc(
-                truth_probabilities, prediction_probabilities, truth_weights.validation_weight
-            )
+        if validation_metric:
```
Comment (brianhelba) [inline isic_challenge_scoring/classification.py:73]: Should update the type annotation / initialization to allow `self.validation` to be optional.

**Comment by brianhelba** [comment]
@danlamanna You'll need to rebase this on `master`. I stole a commit from here to make #110, so you can drop that.

File: isic_challenge_scoring/types.py [hunk 4]
```diff
@@ -9,13 +9,13 @@ class ScoreException(Exception):
 SeriesDict = Dict[str, float]
 DataFrameDict = Dict[str, SeriesDict]
 RocDict = Dict[str, List[float]]
-ScoreDict = Dict[str, Union[float, SeriesDict, DataFrameDict, Dict[str, RocDict]]]
+ScoreDict = Dict[str, Union[Optional[float], SeriesDict, DataFrameDict, Dict[str, RocDict]]]
```
Comment (brianhelba) [inline isic_challenge_scoring/types.py]:
```suggestion
ScoreDict = Dict[str, Union[float, Optional[float], SeriesDict, DataFrameDict, Dict[str, RocDict]]]
```

Unless the type checker doesn't like this, it's more accurate.

File: isic_challenge_scoring/classification.py [hunk 5]
```diff
@@ -61,9 +69,36 @@ def __init__(
...
-            truth_probabilities, prediction_probabilities, truth_weights.validation_weight
-        )
+
+        if validation_metric:
+            if validation_metric == ValidationMetric.BALANCED_ACCURACY:
+                self.validation = metrics.balanced_multiclass_accuracy(
+                    truth_probabilities, prediction_probabilities, truth_weights.validation_weight
+                )
+            elif validation_metric == ValidationMetric.AVERAGE_PRECISION:
+                per_category_ap = pd.DataFrame(
```
Comment (brianhelba) [inline isic_challenge_scoring/classification.py]: Why is this a `DataFrame`? The result of each call to `metrics.average_precision` results in a scalar, so a list of them should just be a `Series`.

File: isic_challenge_scoring/classification.py [hunk 6]
```diff
@@ -61,9 +69,36 @@ def __init__(
...
+                    [
+                        metrics.average_precision(
+                            truth_probabilities[category],
+                            prediction_probabilities[category],
+                            truth_weights.validation_weight,
+                        )
+                        for category in categories
+                    ]
+                )
+                self.validation = per_category_ap.mean(axis='index')
```
Comment (brianhelba) [inline isic_challenge_scoring/classification.py]: Assuming `per_category_ap` ends up as a `Series`, I don't think the additional `axis=` specification should be necessary.

File: isic_challenge_scoring/classification.py [hunk 7]
```diff
@@ -61,9 +69,36 @@ def __init__(
...
+            elif validation_metric == ValidationMetric.AUC:
+                per_category_auc = pd.DataFrame(
```
Comment (brianhelba) [inline isic_challenge_scoring/classification.py]: Same question here as `per_category_ap`.

**Review by brianhelba — CHANGES_REQUESTED** [review]
The new parameters look good. I still have some concerns about the way the metrics are calculated: see inline comments.

I want to add a test for this, which at least exercises the functionality. I can talk about the best ways to do this, when you have time.
